mesonpy/__init__.py

```diff
@@ -953,18 +953,25 @@ def _project(config_settings: Optional[Dict[Any, Any]]) -> Iterator[Project]:
 
 
 def _get_ninja_if_needed() -> List[str]:
-    # only add ninja if it is not present or is too low a version
-    ninja = shutil.which('ninja')
-    if ninja is None:
-        return []
+    env_ninja = os.environ.get('NINJA', None)
+    ninja_canidates = [env_ninja] if env_ninja else ['ninja', 'ninja-build', 'samu']
+    for ninja in ninja_canidates:
```

Comment:
Spelling nit: missing the first "d" in "candidates" (the "candidate_version" variable uses correct spelling below).

Reply:
Thanks! I'll fix this but need to know the best direction to go for the generating project metadata step which seems to happen outside the build hooks. Or does it? It seems rather unlikely it would be doing project generation in the get requires hooks, which should be the only unprotected hooks. Maybe it's fine to add these to the testing requirements, as long as it's also testing inside PEP 517 isolated builds and passing. Is there a dependent project I could quickly try with this branch that's smaller than scipy?

Reply:
I recently offered a bit of review to a project that might be more convenient to test: https://github.com/dfm/python-fsps

Reply:
> for the generating project metadata step which seems to happen outside the build hooks. Or does it? It seems rather unlikely it would be doing project generation in the get requires hooks, which should be the only unprotected hooks.

I'm not entirely sure what all this is doing or *why*, but immediately under the changes you make, it's running a context manager that sets up a build directory, and then checks if "is_pure", and if not pure it goes and adds patchelf.

Reply:
Prefect, I also think I see what I need to do in the tests. It'll take a bit and I'm teaching today, so tonight or tomorrow I'll fix it, add pep 517 tests, and verify that the linked project can build with the change.

Reply:
If someone could retriever the tests, I've added ninja as a testing dependency and added a "PEP 518 test" (which is really a bit more of an integration test for PEP 517 isolation, but pulls in the PEP 518 requirement table too and PEP 518 is short and memorable - I can change the name though if you'd prefer.